sdks/java/harness/src/test/java/org/apache/beam/fn/harness/FnHarnessTest.java

```diff
@@ -118,13 +156,26 @@ public void testLaunchFnHarnessAndTeardownCleanly() throws Exception {
                 .setUrl("localhost:" + controlServer.getPort())
                 .build();
 
-        FnHarness.main("id", options, loggingDescriptor, controlDescriptor);
-        assertThat(instructionResponses, contains(INSTRUCTION_RESPONSE));
+        when(environmentVariableMock.apply("LOGGING_API_SERVICE_DESCRIPTOR"))
+            .thenReturn(TextFormat.printToString(loggingDescriptor));
+        when(environmentVariableMock.apply("CONTROL_API_SERVICE_DESCRIPTOR"))
+            .thenReturn(TextFormat.printToString(controlDescriptor));
+
+        FnHarness.main(environmentVariableMock);
       } finally {
         controlServer.shutdownNow();
       }
     } finally {
       loggingServer.shutdownNow();
     }
+
+    // Verify that we first run onStartup functions before even reading the environment, and that
+    // we then call beforeProcessing functions before executing instructions.
+    InOrder inOrder =
+        inOrder(onStartupMock, beforeProcessingMock, environmentVariableMock, instructionResponses);
+    inOrder.verify(onStartupMock).run();
+    inOrder.verify(environmentVariableMock, atLeastOnce()).apply(any());
+    inOrder.verify(beforeProcessingMock).accept(any());
+    inOrder.verify(instructionResponses).add(INSTRUCTION_RESPONSE);
```

Comment:
I strongly dislike this test. (sorry!)

What is the actual _behavior_ that should change based on the mock. I really don't think a mock is necessary or a good idea here.